[OSDOCS-3974]: new node tuning topic for hosted control planes (#51302)
openshift/openshift-docs: modules/advanced-node-tuning-hosted-cluster.adoc

// Module included in the following assemblies:
//
// * scalability_and_performance/using-node-tuning-operator.adoc

[id="advanced-node-tuning-hosted-cluster_{context}"]
= Advanced node tuning for hosted clusters by setting kernel boot parameters

:FeatureName: Hosted control planes
include::snippets/technology-preview.adoc[]

For more advanced tuning in hosted control planes, which requires setting kernel boot parameters, you can also use the Node Tuning Operator. The following example shows how you can create a node pool with huge pages reserved.

.Procedure

> . Create a config map that contains a tuned object manifest for creating 10 huge pages that are 2 MB in size. Save this config map manifest in a file named `tuned-hugepages.yaml`:

Reviewer: ```suggestion
. Create a `ConfigMap` that contains a `Tuned` object manifest for creating 10 huge pages that are 2 MB in size. Save this `ConfigMap` manifest in a file named `tuned-hugepages.yaml`:
```

+
[source,yaml]
----
    apiVersion: v1
    kind: ConfigMap
    metadata:
      name: tuned-hugepages
      namespace: clusters
    data:

>       tuned: |

Reviewer: ```suggestion
      tuning: |
```

        apiVersion: tuned.openshift.io/v1
        kind: Tuned
        metadata:
          name: hugepages
          namespace: openshift-cluster-node-tuning-operator
        spec:
          profile:
          - data: |
              [main]
              summary=Boot time configuration for hugepages
              include=openshift-node
              [bootloader]
              cmdline_openshift_node_hugepages=hugepagesz=2M hugepages=50
            name: openshift-node-hugepages
          recommend:
          - priority: 20
            profile: openshift-node-hugepages
----
+
[NOTE]
====

> The `.spec.recommend.match` field is intentionally left blank. In this case, this tuned is applied to all nodes in the node pool where this config map is referenced. Group nodes with the same hardware configuration into the same node pool. Otherwise, TuneD operands can calculate conflicting kernel parameters for two or more nodes that share the same node pool.

Reviewer: ```suggestion
The `.spec.recommend.match` field is intentionally left blank. In this case, this `Tuned` is applied to all nodes in the node pool where this `ConfigMap` is referenced. Group nodes with the same hardware configuration into the same node pool. Otherwise, TuneD operands can calculate conflicting kernel parameters for two or more nodes that share the same node pool.
```

====

> . Create the config map in the management cluster:

Reviewer: ```suggestion
. Create the `ConfigMap` in the management cluster:
```

+
[source,terminal]
----
$ oc --kubeconfig="$MGMT_KUBECONFIG" create -f tuned-hugepages.yaml
----

> . Create a node pool manifest YAML file, customize the upgrade type of the node pool, and reference the config map that you created in the `spec.tunedConfig` section. Create the node pool manifest and save it in a file named `hugepages-nodepool.yaml`:

Reviewer: ```suggestion
. Create a `NodePool` manifest YAML file, customize the upgrade type of the `NodePool` and reference the `Config Map` that you created in the `spec.tunedConfig` section. Create the `NodePool` manifest and save it in a file named `hugepages-nodepool.yaml` using the `hypershift` CLI:
```

> . Create a `NodePool` manifest YAML file, customize the upgrade type of the `NodePool`, and reference the `ConfigMap` object that you created in the `spec.tunedConfig` section. Create the `NodePool` manifest and save it in a file named `hugepages-nodepool.yaml` by using the `hypershift` CLI:

Reviewer: ```suggestion
. Create a `NodePool` manifest YAML file, customize the upgrade type of the `NodePool`, and reference the `ConfigMap` object that you created in the `spec.tuningConfig` section. Create the `NodePool` manifest and save it in a file named `hugepages-nodepool.yaml` by using the `hypershift` CLI:
```

+
[source,yaml]
----
    NODEPOOL_NAME=hugepages-example
    INSTANCE_TYPE=m5.2xlarge
    NODEPOOL_REPLICAS=2

    hypershift create nodepool aws \
      --cluster-name $CLUSTER_NAME \
      --name $NODEPOOL_NAME \
      --node-count $NODEPOOL_REPLICAS \
      --instance-type $INSTANCE_TYPE \
      --render > hugepages-nodepool.yaml
----

> . In the `hugepages-nodepool.yaml` file, set `.spec.management.upgradeType` to `InPlace`, and set `.spec.tunedConfig` to reference the `tuned-hugepages` config map that you created.

Reviewer: ```suggestion
. In the `hugepages-nodepool.yaml` file, set `.spec.management.upgradeType` to `InPlace`, and set `.spec.tunedConfig` to reference the `tuned-hugepages` ConfigMap that you created.
```

> . In the `hugepages-nodepool.yaml` file, set `.spec.management.upgradeType` to `InPlace`, and set `.spec.tunedConfig` to reference the `tuned-hugepages` `ConfigMap` object that you created.

Reviewer: ```suggestion
. In the `hugepages-nodepool.yaml` file, set `.spec.management.upgradeType` to `InPlace`, and set `.spec.tuningConfig` to reference the `tuned-hugepages` `ConfigMap` object that you created.
```

+
[source,yaml]
----
    apiVersion: hypershift.openshift.io/v1alpha1
    kind: NodePool
    metadata:
      name: hugepages-nodepool
      namespace: clusters
      ...
    spec:
      management:
        ...
        upgradeType: InPlace
      ...

>       tunedConfig:

Reviewer: ```suggestion
      tuningConfig:
```

      - name: tuned-hugepages
----
+
[NOTE]
====

> To avoid the unnecessary re-creation of nodes when you apply the new 'MachineConfig' objects, set `.spec.management.upgradeType` to `InPlace`. If you use the `Replace` upgrade type, nodes are fully deleted and new nodes can replace them when you apply the new kernel boot parameters that the TunedD operand calculated.

Reviewer: ```suggestion
To avoid the unnecessary re-creation of nodes when you apply the new 'MachineConfig' objects, set `.spec.management.upgradeType` to `InPlace`. If you use the `Replace` upgrade type, nodes are fully deleted and new nodes can replace them when you apply the new kernel boot parameters that the TuneD operand calculated.
```

Reviewer: @lahinson I think this typo is still in the docs.

====

> . Create the node pool in the management cluster:

Reviewer: ```suggestion
. Create the `NodePool` in the management cluster:
```

> To avoid the unnecessary re-creation of nodes when you apply the new 'MachineConfig' objects, set `.spec.management.upgradeType` to `InPlace`. If you use the `Replace` upgrade type, nodes are fully deleted and new nodes can replace them when you apply the new kernel boot parameters that the TuneD operand calculated.

Author: Good catch, @dagrayvid! I'll be sure to fix those.

+
[source,terminal]
----
$ oc --kubeconfig="$MGMT_KUBECONFIG" create -f hugepages-nodepool.yaml
----

.Verification

After the nodes are available, the containerized TuneD daemon calculates the required kernel boot parameters based on the applied tuned profile. After the nodes are ready and reboot once to apply the generated 'MachineConfig' object, you can verify that the tuned profile is applied and that the kernel boot parameters are set.

> . List the tuned objects in the hosted cluster:

Reviewer: ```suggestion
. List the `Tuned` objects in the hosted cluster:
```

> After the nodes are available, the containerized TuneD daemon calculates the required kernel boot parameters based on the applied TuneD profile. After the nodes are ready and reboot once to apply the generated 'MachineConfig' object, you can verify that the TuneD profile is applied and that the kernel boot parameters are set.

Reviewer: ```suggestion
After the nodes are available, the containerized TuneD daemon calculates the required kernel boot parameters based on the applied TuneD profile. After the nodes are ready and reboot once to apply the generated `MachineConfig` object, you can verify that the TuneD profile is applied and that the kernel boot parameters are set.
```

+
[source,terminal]
----
$ oc --kubeconfig="$HC_KUBECONFIG" get Tuneds -n openshift-cluster-node-tuning-operator
----
+
.Example output
[source,terminal]
----
NAME                 AGE
default              123m
hugepages-8dfb1fed   1m23s
rendered             123m
----
   

> . List the profiles in the hosted cluster:

Reviewer: ```suggestion
. List the `Profile` objects in the hosted cluster:
```

+
[source,terminal]
----
$ oc --kubeconfig="$HC_KUBECONFIG" get Profiles -n openshift-cluster-node-tuning-operator
----
+
.Example output
[source,terminal]
----
NAME                           TUNED                      APPLIED   DEGRADED   AGE
nodepool-1-worker-1            openshift-node             True      False      132m
nodepool-1-worker-2            openshift-node             True      False      131m
hugepages-nodepool-worker-1    openshift-node-hugepages   True      False      4m8s
hugepages-nodepool-worker-2    openshift-node-hugepages   True      False      3m57s
----
+

> Both of the worker nodes in the new node pool have the `openshift-node-hugepages` profile applied.

Reviewer: ```suggestion
Both of the worker nodes in the new `NodePool` have the `openshift-node-hugepages` profile applied.
```
